File: cpp/leetcode/125.cpp
#include <vector>
#include <iostream>

using namespace std;

bool checkPalindrome(string str) {
    //preprocessing the string
    string modifiedStr = "";
    for (char ch : str) {
        if (isalnum(ch) == true) {
            char modifiedChar = tolower(ch);
            modifiedStr.push_back(modifiedChar);
        }
    }

    for (int i = 0; i < modifiedStr.size() / 2; i++) {
        if (modifiedStr[i] != modifiedStr[modifiedStr.size() - 1 - i]) {
            return false;
        }
    }

    return true;

}

int main() {

    string input = "a man A plAn a cAnaL : panama!";
    string input2 = "race a car";

    cout << checkPalindrome(input) << endl;;
    cout << checkPalindrome(input2) << endl;;

    

    return 0;
}

File: python/grind75/125.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        filtered_arr = []
        for ch in s:
            if ch.isalnum():
                filtered_arr.append(ch.lower())

        return filtered_arr == filtered_arr[::-1]


if __name__ == "__main__":
    # s = "A man, a plan, a canal: Panama"
    s = "0P"
    solution = Solution()
    print(solution.isPalindrome(s))
